### PR TITLE
*: introduce pkg/migrations

### PR DESCRIPTION
**Pull request by irfansharif**
Implements the long-running migrations RFC (#48843). This PR builds on top a
few core components we've introduced to crdb over the past few months. They
are (in order):

- #45455 (re-works cluster version settings to be a standalone component)
- #50329 (introduces a fully decommissioned bit)
- #52526 (introduces the join RPC for node ID allocation, and to hand out
          active cluster versions to joining nodes)
- #53842 (always creating a liveness record for nodes, before startup)
- #55286 (blocks decommissioned nodes from interacting with the cluster)

The components above (save for #55286) were included as part of the 20.2
release. That paved the way for us to introduce this machinery for 21.1 (it
also lets us define migrations starting this cycle). For the detailed story
around how it all fits together, read the RFC.

---

Big picture: going forward `SET CLUSTER SETTING version = <whatever>` will step
through each intermediate (unstable) version between what's currently active,
and what the intended target is, and execute associated migrations for each
one (if any). For each "step", the migration manager first checks to see that
every node in the cluster is running an appropriate binary version. It then
activates the relevant version gate on every node (through a new RPC, instead
of gossip like we used to previously). Finally it executes the migration
itself, providing it with the guarantee that all `.IsActive(VersionTag)` will
return `true` going forward. After successfully migrating through each
intermediate version, the statement returns.

---

Package migration captures the facilities needed to define and execute
migrations for a crdb cluster. These migrations can be long running, are free
to send out arbitrary requests cluster wide, change internal DB state, and much
more. They're typically reserved for crdb internal operations and state. Each
migration is idempotent in nature, is associated with a specific cluster
version, and executed when the cluster version is made activate on every node
in the cluster.

Examples of migrations that apply would be migrations to move all raft
state from one storage engine to another, or purging all usage of the
replicated truncated state in KV. 

---

To get a sense of the primitives made available to migrations, a rough sketch
of the APIs is presented below.

```go
  type Migration func(context.Context, *Helper) error
  func ExampleMigration (ctx context.Context, h *Helper) error {
    // NB: The harness has already made sure that the version associated with
    // this particular migration this is already active on all nodes in the
    // cluster.

    // Migrations have access to a *kv.DB.
    txn := h.DB().NewTxn(ctx, ...)

    // Migrations can run for arbitrarily long.
    time.Sleep(time.Second)

    // Migrations can run SQL.
    _, err := h.Conn().Queryf(`SELECT 1`)

    // Migrations can paginate through each range descriptor.
    _ = h.IterRangeDescriptors(pageSize, func(descriptors ...roachpb.RangeDescriptor) error {
      for _, desc := range descriptors {
        // Migrations can selective execute below-raft migrations.
        h.DB().Migrate(ctx, desc.StartKey, desc.EndKey, h.Version())
      }

      // Migrations can persist progress to a system table, for introspection.
      h.Logf("paginated through %d descriptors", len(descriptors))
    })

    // Migrations can define/execute server level operations across every node
    // (like forcing all replicas through GC, etc.)
    h.EveryNode(ctx, op)
  }
```

---

See individual commits. This PR introduces individual components without it
being wired up to anything first, and then slowly incorporates them into
actual running code. The commits of note (in order) are pulled out below.

```
  sql: add scaffolding for version upgrade hook
  
  This callback will be called after validating a `SET CLUSTER SETTING
  version` but before executing it. It will be used in future commits to
  execute long-running migrations.
```

```
  server,kvserver: introduce the `EveryNode` rpc
  
  ...and populate it with the `AckClusterVersion` op.
  
  EveryNode is the RPC that will powers the long running migrations
  infrastructure. It'll let callers define and execute arbitrary commands
  across every node in the cluster. To motivate what this would look like,
  we introduce alongside it one such command, the `AckClusterVersion`
  operation. It isn't currently hooked up to anything, but this will
  eventually be the mechanism through which we'll propagate cluster
  version bumps across the crdb cluster, replacing our gossip based
  distribution mechanism in-place today. This will let the orchestrator
  bump version gates in a controlled fashion across the cluster.
  
  To achieve this, the stubbed out implementation of AckClusterVersion
  makes use of the same `StoreClusterVersionKey` otherwise used in
  callbacks attached to the gossip handler for cluster version bumps.
```

```
  kvserver: introduce GetLivenessesFromKV
  
  Now that we always create a liveness record on start up (#53805), we can
  simply fetch all liveness records from KV when wanting an up-to-date
  view of all nodes in the cluster. We add a helper to do as much,
  which we'll rely on in future commits. 

  It's a bit unfortunate that we're further adding on to the NodeLiveness API
  without changing the underlying look-aside caching structure, but the
  methods fetching records from KV is the world we're hoping to start moving
  towards over time.
```

```
  migration: introduce pkg/migrations
  
  Package migration captures the facilities needed to define and execute
  migrations for a crdb cluster. These migrations can be arbitrarily long
  running, are free to send out arbitrary requests cluster wide, change
  internal DB state, and much more. They're typically reserved for crdb
  internal operations and state. Each migration is idempotent in nature,
  is associated with a specific cluster version, and executed when the
  cluster version is made activate on every node in the cluster.
  
  Examples of migrations that apply would be migrations to move all raft
  state from one storage engine to another, or purging all usage of the
  replicated truncated state in KV. A "sister" package of interest is
  pkg/sqlmigrations.
  
  This commit only introduces the basic scaffolding and wiring from
  existing functionality. We'll flesh in the missing bits in future
  commits.
```

```
  server,migration: generate EveryNode req/resp helpers
  
  We expect to add multiple req/resp types as individual operations for
  the EveryNode RPC. Each of these operations will correspond to a
  "primitive" of sorts for the (long running) migrations infrastructure.
  It's a bit cumbersome to wield this nested union type, so we
  autogenerate helper code to do it for us. We take precedence from
  api.proto and all the very many batch request/responses.
```    

```
  server: expose GetLivenessesFromKV in node liveness interfaces
  
  To flesh out the migrations infrastructure (in future commits), we'll
  need a handle on all the nodes present in the system. Now that we always
  create a liveness record on start up (#53805), we can simply fetch all
  liveness records from KV. We add a helper to do so.
  
  This does mean that we're introducing a direct dependency on
  NodeLiveness in the sql layer, and there's improvements to be made here
  around interfaces delineating between "system tenant only" sql code and
  everything else. Only system tenants have the privilege to set cluster
  settings (or at least the version setting specifically), which is what
  this API will look to power.
```   

```
  server: introduce ValidateTargetClusterVersionRequest
  
  We'll use this primitive in a future commit to introduce additional
  safeguards (not present today) around cluster version upgrades.
  Specifically, we'll use this EveryNode operation to validate that every
  node in the cluster is running a binary that's able to support the
  specified cluster version.
```

```
  settings,clusterversion: disconnect cluster version from gossip
  
  ...in favor of direct RPCs to all nodes in the cluster. This commit in
  particular deserves a bit of scrutiny. It uses the building blocks we've
  added thus far to replace the use of gossip to disseminate the cluster
  version. It does so by sending out individual RPCs to each node in the
  cluster, informing them of a version bump, all the while retaining the
  same guarantees provided by our (now previously) gossip-backed mechanism.
  
  This diff has the following "pieces":
  - It disconnecting version setting updates through gossip (by
    disconnecting the setting type within the updater process)
  - It using the EveryNode RPC to send out RPCs to each node in the
    cluster, containing the payload that each node would otherwise receive
    through gossip.
  - It expands the clusterversion.Handle interface to allow setting the
    active version directly through it.
  - It persisting any cluster versions received from other nodes first,
    within keys.StoreClusterVersionKey, before bumping the version gate.
    This was previously achieved by attaching callbacks on the version
    handle (look towards all the diffs around SetBeforeChange). This is an
    invariant we also maintained earlier.
  - It using the active version provided by the join RPC to set the
    version setting directly (after having persisted it first). This too
    was previously achieved through gossip + the callback.
  
  While here, we add a few comments and chip away at the callback hooks
  that are no longer needed.
```

```
  sql: introduce system.migrations
  
  We'll use it in a future commit to store migration state, for
  introspection.
  
  > SHOW CREATE system.migrations;
           table_name        |                     create_statement
  ---------------------------+------------------------------------------------------------
    system.public.migrations | CREATE TABLE public.migrations (
                             |     id INT8 NOT NULL DEFAULT unique_rowid(),
                             |     metadata STRING NOT NULL,
                             |     started TIMESTAMP NOT NULL DEFAULT now():::TIMESTAMP,
                             |     progress BYTES NULL,
                             |     CONSTRAINT "primary" PRIMARY KEY (id ASC),
                             |     FAMILY "primary" (id, metadata, started),
                             |     FAMILY progress (progress)
                             | )
```

---

Still need to add tests. But it "works".

### CODE REVIEW

File: pkg/server/server_sql.go [hunk 1]
```diff
@@ -770,7 +770,7 @@ func (s *sqlServer) preStart(
 		return errors.Wrap(err, "ensuring SQL migrations")
 	}
 
-	log.Infof(ctx, "done ensuring all necessary migrations have run")
+	log.Infof(ctx, "done ensuring all necessary startup migrations have run")
```
Comment (tbg) [inline pkg/server/server_sql.go:797]: :+1: I'd say throw this one into a separate PR, as this one won't merge for a while and that way it's out of your hair

File: pkg/kv/batch.go [hunk 2]
```diff
@@ -754,3 +754,27 @@ func (b *Batch) addSSTable(
 	b.appendReqs(req)
 	b.initResult(1, 0, notRaw, nil)
 }
+
+// migrate is only exported on DB.
+//lint:ignore U1001 unused
+func (b *Batch) migrate(s, e interface{}, targetVersion roachpb.Version) {
```
Comment (tbg) [inline pkg/kv/batch.go:760]: This commit belongs into a separate PR that revamps https://github.com/cockroachdb/cockroach/pull/42822

File: pkg/server/serverpb/admin.proto [hunk 3]
```diff
@@ -455,6 +456,57 @@ message DecommissionStatusResponse {
...
+// EveryNodeRequestUnion. These requests will power the long running migrations
+// process where the orchestrator of these long running migrations may need to
+// run arbitrary operations across every node in the cluster (depending on the
+// specific migration being executed). One such example is the orchestrator
+// wanting to inform every node in the cluster that a certain cluster version is
+// now active, and that each node should enable the relevant version gates
+// corresponding to said version.
+//
+// TODO(irfansharif): Actually use these as outlined above.
+message EveryNodeRequest {
```
Comment (tbg) [inline pkg/server/serverpb/admin.proto:517]: As we briefly discussed, `EveryNode` is not a fitting name. That name somehow suggests that this RPC internally will do something on every node, but this is not how it works.

I also come down on the side of thinking that this should be a service and that an RPC can be added for each individual task. The argument that I've heard for the status quo in this PR, that it is somehow less work, does not convince me as you have also added custom codegen, so it's `make generate` vs `make protobuf`. At that point I would rather stick with the idiomatic choice:

```go
service Migrations {
  rpc BumpClusterVersion(BumpClusterVersionRequest) returns BumpClusterVersionResponse {}
  rpc CheckReplicaVersions(CheckReplicaVersionsRequest) returns CheckReplicaVersionsResponse {} // maybe needed for range migrations
  // ...
}
```

I also think this is a good work item to go into a separate PR. It's 100% clear that we will need it, and until the comments are just right etc I think we'll have 1-2 more review iterations that would clutter up this main PR.

File: pkg/kv/kvserver/node_liveness.go [hunk 4]
```diff
@@ -128,6 +128,8 @@ type LivenessMetrics struct {
 	HeartbeatLatency   *metric.Histogram
 }
 
+// TODO(irfansharif): We should/could cut out a pkg/nodeliveness here.
```
Comment (tbg) [inline pkg/kv/kvserver/node_liveness.go:131]: If you think doing this would make you really happy, then I would support doing it now. Otherwise, and this is my intuition about what's best right now, remove the comment and leave as is. You're not materially changing liveness here, and we know liveness is a clear candidate to move to a pkg.

Can you spin this commit out into a separate PR? This can then be landed already, once you've added a simple unit test.

File: pkg/kv/kvserver/node_liveness.go [hunk 5]
```diff
@@ -1002,6 +1004,39 @@ func (nl *NodeLiveness) GetLivenesses() []kvserverpb.Liveness {
 	return livenesses
 }
 
+// GetLivenessesFromKV returns a slice containing the liveness status of every
+// node on the cluster. It's the improved version of GetLivenesses above, which
```
Comment (tbg) [inline pkg/kv/kvserver/node_liveness.go:1008]: it's not "improved", it serves a different purpose. If being cheap is what you're after, obviously you'd avoid the KV version. We are adding this because we need it for correctness, so a comment could say

> GetLivenessesFromKV returns a slice containing the liveness record of all nodes that have ever been a part of the cluster. The records are read from the KV layer in a KV transaction.

We also need to consider whether this needs to paginate and possibly avoid old node IDs. I hope it does not (i.e. back of the envelope calculations should show that we'll never crack a couple of mbs here).

File: pkg/roachpb/metadata.proto [hunk 6]
```diff
@@ -382,5 +382,9 @@ message Version {
   // The unstable version is used to migrate during development.
   // Users of stable, public releases will only use binaries
   // with unstable set to 0.
+  //
+  // TODO(irfansharif): We'll now use `unstable` to tack on long running
+  // migrations. Should we rename this to `intermediate` or something similar?
+  // We should at least motivate the purpose of this field a bit better.
   optional int32 unstable = 4 [(gogoproto.nullable) = false];
```
Comment (tbg) [inline pkg/roachpb/metadata.proto:389]: the name "internal" seems fitting. They are subversions that are never the end versions of a release.

Mind sending out a separate PR that makes this change if you like that name?

File: pkg/migration/manager.go [hunk 7]
```diff
@@ -109,9 +110,34 @@ func NewManager(
 //
 // TODO(irfansharif): We want to eventually be able to send out a Migrate
 // request spanning the entire keyspace. We'll need to make sure all stores have
-// synced once to persist any raft command applications. Implement this.
+// synced once to persist any raft command applications.
 //lint:ignore U1001 unused
-func (h *Helper) IterateRangeDescriptors(f func(...roachpb.RangeDescriptor) error) error {
+func (h *Helper) IterateRangeDescriptors(
```
Comment (tbg) [inline pkg/migration/manager.go:115]: Move (and postpone) this whole commit please. Let's keep the helper as small as possible for now and add these gadgets in later when we have real migrations to work with.

File: pkg/sql/catalog/systemschema/system.go [hunk 8]
```diff
@@ -342,6 +342,17 @@ CREATE TABLE system.sqlliveness (
     expiration       DECIMAL NOT NULL,
   	FAMILY fam0_session_id_expiration (session_id, expiration)
 )`
+
+	MigrationsTableSchema = `
```
Comment (tbg) [inline pkg/sql/catalog/systemschema/system.go:346]: Upon reflecting this a bit, I think we should take a step back. The main goals here are

a) allow scripting against the migration process, i.e. allow operators to determine what the migration status is
b) allow troubleshooting when a migration goes wrong

We're unlikely to be able to achieve b) using just a SQL table. However, for a) a SQL table is a good fit. So let's optimize this table for A and open an issue about B.

The table can thus become relatively simple. I think there should be one row per version that does work (and one for major versions) with timing information, name, and a status that the migration work function can update at will. For example an active row could look like this

```go
"20.2-13", "running", "VersionFooBar", "fooing all the bars", <start ts>, <end ts if finished>, <last error if previous attempt failed>
```

My suggestion is to split this out into a separate PR (the table is not needed for the basic harness and the basic version of the manager and testing it). I would be curious what @ajwerner thinks of our schema attempts here.

for b), I wonder if we can leverage loggers. Basically, we give the work function a logger that logs to an auxiliary file, and make it clear that this logger should be used liberally. Here I am interested in @knz's input.
Additionally, we could also open a net/trace span and hook that up to the context, but this will be less useful as we don't collect net/trace in debug.zip. But - it could be a nice auxiliary way to expose it that can come in handy on systems where access to logs is difficult to attain.

File: pkg/migration/manager.go [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+	return nodeIDs, nil
+}
+
+// Log lets migrations insert arbitrary log events. These events are recorded in
+// a system table, and are available for introspection into the progress of the
+// specific ongoing migration.
+//
+// TODO(irfansharif): Implement this.
+//lint:ignore U1001 unused
+func (h *Helper) Log(event string) error {
```
Comment (tbg) [inline pkg/migration/manager.go:179]: Let's strip everything that isn't essential for now.

File: pkg/migration/manager.go [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+			op := &serverpb.AckClusterVersionRequest{Version: &version}
+			if err := h.EveryNode(ctx, op); err != nil {
+				return err
+			}
+		}
+
+		// TODO(irfansharif): We'll want a testing override here to be able to
+		// stub out migrations as needed.
+
+		migration, ok := Registry[version]
```
Comment (tbg) [inline pkg/migration/manager.go:273]: We also need to run a number of migrations, not just one, right?

Comment (tbg) [reply pkg/migration/manager.go:273]: The code I was expecting here was something like this

```go
for _, v := range clusterversion.GetVersionsFromRange(initialVersion, targetVersion) {
   migration, ok := Registry[v]
   if !ok { continue }
   migration(ctx, &Helper{...})
}
```

Comment (tbg) [reply pkg/migration/manager.go:273]: Also, we need "fence versions". We can't run the migration into, say, 21.1-15 before all nodes have acknowleged the fact that a migration is ongoing. So basically each version with work attached to it needs to have an "empty" version before it. Example:

Instead of
```
20.2-1 migrate lock table
20.2-2 migrate raft log
```

It would have to be
```
20.1-1 noop
20.1-2 migrate lock table
20.1-3 noop
20.1-4 migrate raft log
```

Otherwise, you could get into situations in which 20.1-4 is not active yet, but some nodes have already started migrating their lock tables. Maybe this is fine, but in particular the first migration in a version is critical. Without the noop, we might start migrating lock tables before bumping the clusterversion up from `20.1`, which means the operator can roll back to a 20.1 binary that doesn't understand the migrated lock table.
I think for now we can just add an init function that does some validation of the `Registry` to make sure that there's always a noop bump before any work bump. Later, if this is too annoying a papercut, we can change the way clusterversions are defined so that we assign the `unstable` component automatically, and so we'll just use `2*i+1` instead of `i` in the current enum.

File: pkg/migration/manager.go [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+		// return. The migration associated with the specific version can assume
+		// that every node in the cluster has the corresponding version
+		// activated.
+		{
+			// First sanity check that we'll actually be able to perform the
+			// cluster version bump, cluster-wide.
+			//
+			// TODO(irfansharif): Do we want to worry about the case where we
+			// validate N nodes running the right binary versions, and the user
+			// adding another node that would no longer be able to? As written,
```
Comment (tbg) [inline pkg/migration/manager.go:254]: Don't quite understand, you mean the case where a node joins after the validation but before the bump?

I thought we discussed this during the RFC. We need to roll out the cluster version (at least) twice. When we've done it once, we know that all the nodes we saw the first time around have the updated version, so they will only let new nodes join at that version. But some could've snuck in, but that's ok - we get the list of nodes again, do it one more time, and see if we caught any new nodes. If we didn't, done. If we did, do it again until it stabilizes (as that new node might've let yet another new node slip in).

**Comment by tbg** [comment]
Nice! Thanks for pre-polishing these commits. I spent quite a bit of time looking at this and thinking about how to most efficiently break it down. There is too much work here, and everything does need a bit of polish and consideration, and we won't get the best outcome if we try doing that all in here.

What I would suggest is to send out separate PRs:

```
# separate PR, easy merge
pick 711245ad46 server: clarify a logging message

# separate PR, just needs a little test, then good to go
pick c3801a56f2 kvserver: introduce GetLivenessesFromKV
pick 829697b340 server: expose GetLivenessesFromKV in node liveness interfaces

# separate PR, probably a bit more bikeshedding and polish needed, should add a first RPC (VerifyVersion?)
pick e27583aa57 kvserver,roachpb: introduce Migrate request type
pick c690d2db3b server,kvserver: introduce the `EveryNode` rpc
```

The above are nice to get out of the way because while there's a bit of code yet to write and review to be had, there's not a lot that's tricky about them.

Then:

```
# separate PR: introduce version hook, add noop migration manager, call it before version bump 
pick aad0bad3ac sql: add scaffolding for version upgrade hook 
pick 32397d6963 migration: introduce pkg/migrations 
pick 9ccb3f9627 migration: plumb in node dialer 
pick a43a6bfb7a migration: plumb in an internal executor, kv.DB 
 ```

This gets most of the "boring code" out of the way. The manager is hooked up, in the right place, but doesn't do anything.

Then two items that I think aren't in any commit yet (the first refers mostly to https://github.com/cockroachdb/cockroach/pull/56107#discussion_r515097576). For testing this, I expect that we'll find that we want to trim down the dependencies to the manager because we want to test its validation logic in isolation. That is, we want to test the manager's behavior without a surrounding server. For example, we wouldn't hand it a `*NodeLiveness` but instead a getter function. And we would do something similar for any other external dependency that it has so that we can simulate the "nodes joining at inopportune times", etc, cases as true unit tests, and can also give the manager a few migrations and make sure it steps through them in the right order.

```
# missing: 
pick 0000000000 PR that makes the migration manager correct and tested 
pick 0000000000 init(): ensure noop versions before any versions with migrations 
```

After that point, it's time for the real forklift, which I think is mostly

```
# revisit this separately at the end of current merging bonanza, this then 
# actually hooks up the manager to cluster versions 
pick 87c6d56452 settings,clusterversion: disconnect cluster version from gossip
```

When we have that in, we basically "have" long-running migrations, except that we don't have the right observability yet, and we haven't onboarded a real migration. So there is work to figure out the SQL table, figure out auxiliary logging, figure out which migration to onboard, etc.

Let me know what you think. Probably sync is best, but if you don't catch me today Tuesday is the earliest opportunity after.

**Comment by knz** [comment]
I concur that I would prefer to review this piecemeal.

Also a question that I couldn't get answered directly from previous discussions: what if the migration function returns an error? Does it get re-run later?

Also I think the explanation above should explain / insist that the person implementing migrations should ensure they are idempotent.

File: pkg/server/server_sql.go (hunk 1 above)
Comment (irfansharif) [reply pkg/server/server_sql.go:797]: https://github.com/cockroachdb/cockroach/pull/56194 (I never know with small diffs like this, they're not cumbersome enough for me to want to pull them out separately).

File: pkg/roachpb/metadata.proto (hunk 6 above)
Comment (irfansharif) [reply pkg/roachpb/metadata.proto:389]: Done, https://github.com/cockroachdb/cockroach/pull/56200.

File: pkg/kv/kvserver/node_liveness.go (hunk 4 above)
Comment (irfansharif) [reply pkg/kv/kvserver/node_liveness.go:131]: re: pkg/liveness, https://github.com/cockroachdb/cockroach/pull/56221. I'll send a follow-on PR for GetLivenessFromKV.

Comment (irfansharif) [reply pkg/kv/kvserver/node_liveness.go:131]: > I'll send a follow-on PR for GetLivenessesFromKV.

Done: https://github.com/cockroachdb/cockroach/pull/56243.

File: pkg/kv/kvserver/node_liveness.go (hunk 5 above)
Comment (irfansharif) [reply pkg/kv/kvserver/node_liveness.go:1008]: yup, we shouldn't need to paginate here. Updated this comment in #56243.

File: pkg/migration/manager.go (hunk 11 above)
Comment (irfansharif) [reply pkg/migration/manager.go:254]: > you mean the case where a node joins after the validation but before the bump?

Yes, but I wasn't even thinking about the safety of the migration, I'm thinking about the safety of the version bump itself. Specifically the hazard I'm worried about is the following:
- N nodes running v21.1
- N nodes rolled into v21.2 binaries, but with active cluster version still as v21.1
- Validation for setting active cluster version to v21.2 checks to see that all nodes are running v21.2 binaries
- A new node is added to the cluster, but running binary v21.1
- We start bumping the cluster gates to v21.2, despite there now being a v21.1 node in the cluster.

To be clear, this isn't something we protect against today. Even with this dedicated ValidateTargetClusterVersion rpc, we'll not be fully protecting against it, it'll just be best effort. So perhaps that's ok.

File: pkg/migration/manager.go (hunk 10 above)
Comment (irfansharif) [reply pkg/migration/manager.go:273]: #57186.

File: pkg/migration/manager.go (hunk 11 above)
Comment (irfansharif) [reply pkg/migration/manager.go:254]: Added fence versions in #57186, which handles the hazards above.

File: pkg/sql/catalog/systemschema/system.go (hunk 8 above)
Comment (irfansharif) [reply pkg/sql/catalog/systemschema/system.go:346]: #57445, let's carry the discussion over there.

File: pkg/migration/manager.go (hunk 9 above)
Comment (irfansharif) [reply pkg/migration/manager.go:179]: A version of this is now implemented in #57445.

File: pkg/migration/manager.go (hunk 7 above)
Comment (irfansharif) [reply pkg/migration/manager.go:115]: Tacked along with a real migration in #57445.

File: pkg/kv/batch.go (hunk 2 above)
Comment (irfansharif) [reply pkg/kv/batch.go:760]: #57445.

**Comment by irfansharif** [comment]
This prototype was shredded up into all the various PRs that have since landed on master (see the referring commits above). The few comments here were addressed in those other PRs. The last in-flight one is #57445, which moves past what was prototyped here, so closing this one out. Thanks for the reviews!
